pkg/controllers/internalmembercluster/v1beta1/member_controller.go

```diff
@@ -158,15 +247,172 @@ func (r *Reconciler) stopAgents(ctx context.Context, imc *clusterv1beta1.Interna
+		klog.ErrorS(err, "failed to report cluster properties using built-in mechanism", "InternalMemberCluster", klog.KObj(imc))
+		return err
+	}
+	return nil
+}
+
+// reportPropertyProviderCollectionCondition reports the condition of whether a properity
+// collection attempt has been successful.
+func reportPropertyProviderCollectionCondition(imc *clusterv1beta1.InternalMemberCluster, status metav1.ConditionStatus, reason, message string) {
+	meta.SetStatusCondition(&imc.Status.Conditions, metav1.Condition{
+		Type:               string(clusterv1beta1.ConditionTypeClusterPropertyCollectionSucceeded),
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		ObservedGeneration: imc.GetGeneration(),
+	})
+}
+
+// reportPropertyProviderStartedCondition reports the condition of whether a property provider
+// has been started successfully.
+func reportPropertyProviderStartedCondition(imc *clusterv1beta1.InternalMemberCluster, status metav1.ConditionStatus, reason, message string) {
+	meta.SetStatusCondition(&imc.Status.Conditions, metav1.Condition{
+		Type:               string(clusterv1beta1.ConditionTypeClusterPropertyProviderStarted),
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		ObservedGeneration: imc.GetGeneration(),
+	})
+}
+
+// reportClusterPropertiesWithPropertyProvider collects the latest cluster properties from the property provider.
+func (r *Reconciler) reportClusterPropertiesWithPropertyProvider(ctx context.Context, imc *clusterv1beta1.InternalMemberCluster) error {
+	// Check if there are queued calls that have not returned yet.
+	ticket := r.queuedPropertyCollectionCalls.Add(1)
+	defer r.queuedPropertyCollectionCalls.Add(-1)
+	if ticket > int32(maxQueuedPropertyCollectionCalls) {
+		// There are too many on-going calls to the property provider; report the failure
+		// and skip this collection attempt.
+		reportPropertyProviderCollectionCondition(imc,
+			metav1.ConditionFalse,
+			ClusterPropertyCollectionFailedTooManyCallsReason,
+			ClusterPropertyCollectionFailedTooManyCallsMessage,
+		)
+		err := fmt.Errorf("too many on-going calls to the property provider; skipping this collection attempt")
+		klog.ErrorS(err, "failed to collect cluster properties", "InternalMemberCluster", klog.KObj(imc))
+		return err
+	}
+
+	// Collect the latest cluster properties from the property provider.
+	//
+	// If the property provider fails to return the latest cluster properties within the deadline,
+	// report the failure as a condition.
+	childCtx, childCancelFunc := context.WithDeadline(ctx, time.Now().Add(propertyProviderCollectionDeadline))
+	// Cancel the context any way to avoid leaks.
+	defer childCancelFunc()
+	collectedCh := make(chan struct{})
+
+	var res propertyprovider.PropertyCollectionResponse
+	go func() {
+		defer close(collectedCh)
+		res = r.propertyProvider.Collect(childCtx)
+	}()
+
+	select {
```

Comment:
I am not 100% sure if I understand the question correctly, but the workqueue has guaranteed that for any individual key enqueued it will not be processed concurrently; in addition, there is only one InternalMemberCluster object visible to the member agent any way.

And the behavior is as described: it will either wait until a response is returned in time or time out (and fall back to the default mechanism). The processing has already started before the `select` statement runs.